src/n-dhcp4-c-connection.c

```diff
@@ -1146,6 +1146,7 @@ int n_dhcp4_c_connection_dispatch_timer(NDhcp4CConnection *connection,
 int n_dhcp4_c_connection_dispatch_io(NDhcp4CConnection *connection,
                                      NDhcp4Incoming **messagep) {
         _c_cleanup_(n_dhcp4_incoming_freep) NDhcp4Incoming *message = NULL;
+        static c_thread_local uint8_t scratch_buffer[UINT16_MAX];
```

Comment:
You suddenly increase the memory usage of any n-dhcp4 user that has many threads but only uses a single dhcp4 probe. I think this is ok'ish, but should be considered when changing this. Before this change, this was linear with the number of probes, now it is in no way bound to the usage of the library but rather a property of the parent application.

Reply:
Secondly, you now return a pointer to a `thread_local` from this function, without limiting what the caller can do with it. This is scary and easy to get wrong. Granted, before this change we were also lifetime-bound to `connection`, but that was IMO less fragile.

Anyway, I think the correct fix is to make the only caller of this function (i.e. `n_dhcp4_client_probe_dispatch_io()`) pass in the buffer, just like we pass it to `n_dhcp4_c_socket_packet_recv()` and friends. This way, the caller controls the allocation and can be sure to never reveal any information to the outside.

Lastly, wouldn't it be easier and equally fine to just malloc() the buffer there, and then free it when done? Is this thread-local really a significant performance optimization that we want?

Reply:
> You suddenly increase the memory usage of any n-dhcp4 user that has many threads

Good point. No concern for NetworkManager (which has few threads and many client instances), but might be a problem for others.

> Secondly, you now return a pointer to a thread_local from this function

I don't see that the buffer gets returned to the caller. Could you point that out?

Reply:
we could also do:

```
static c_thread_local uint8_t **scratch_buffer;
const unsigned SCRATCH_BUFFER_SIZE = UINT16_MAX;

...

if (_c_unlikely_(!scratch_buffer)) {
    scratch_buffer = malloc(SCRATCH_BUFFER_SIZE);
    if (!scratch_buffer)
        return -ENOMEM;
}
```
then the wasted memory per thread would be *only* a pointer, while there would be no extra malloc/free per recvmsg.

Reply:
ah, never mind. Of course that would leak the buffers when the thread exists... which we would have to free with some pthread cleanup function...

Reply:
> I don't see that the buffer gets returned to the caller. Could you point that out?

Oops, we copy the data out into the message. I thought we merely reference it from the message. Sorry for the confusion.

Reply:
> ah, never mind. Of course that would leak the buffers when the thread exists... which we would have to free with some pthread cleanup function...

I wouldn't mind checking out some more complex solutions IF they are necessary. But I really doubt the malloc/free would matter much :-/

Reply:
Sorry for the late reply.
I agree, one allocation/free for every received packet should be fine. I'll update the MR